src/main/java/com/google/collegeplanner/servlets/DatastoreServlet.java

```diff
@@ -138,29 +141,52 @@ private void addCourses(JSONArray coursesArray, HttpServletResponse response) th
         return;
       }
       JSONArray sectionsArray = apiUtil.getJsonArray(uri);
-      addSectionsToCourse(sectionsArray, course, courseEntity);
+      addSectionsToCourse(sectionsArray, course, courseEntity, response);
     }
   }
 
   /**
-   * Adds the Course and Section to datastore.
+   * Adds the Course, Sections, and Meetings to datastore.
    * @param sectionsArray The section json from the UMD API.
    * @param course The Course object.
    * @param courseEntity The Entity object that we want to add to datastore.
+   * @param response The HttpServletResponse object.
    */
-  private void addSectionsToCourse(JSONArray sectionsArray, Course course, Entity courseEntity) {
-    if (sectionsArray == null || course == null || courseEntity == null) {
+  private void addSectionsToCourse(JSONArray sectionsArray, Course course, Entity courseEntity,
+      HttpServletResponse response) throws IOException {
+    if (sectionsArray == null || course == null || courseEntity == null || response == null) {
+      respondWithError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, response);
```

Comment:
responding with an error and then returning here can cause some issues. The caller function doesn't really know that anything went wrong.

I think there's two ways to fix this:

1. Return a boolean from this function (true if it was successful), and then propagate that up to doPost, which can set the error properly.
2. Throw an exception and have doPost catch that exception, which then returns the error properly.

Reply:
I'm actually thinking about throwing an error if one of the arguments is null, but then just `continue`-ing for parse exceptions. I don't think the whole program should stop because of a parse exception on just one course. I think it should still add the rest of the courses that aren't messed up.

Reply:
Yep, that makes sense. In that case, you probably would return a 200 then instead of returning 500 whenever any section fails. What you can do is also report the status in the response (X number of courses imported, Y failed, etc).

Reply:
After `continue`-ing it should eventually reach the statement where I return status 200 in `doPost()`. Do you think returning how many succeeded is necessary?

Reply:
How will you know if anything failed?